File: bot.py
# todo
	# add readme on github
	# check if conditions are proper in game_over()


#region imports
import random
import math
#endregion


#region function definitions


# game_over checks whether the game has ended and returns the winner
def game_over(board):
	if board[1] == board[2] == board[3]:
		winner = board[1]
		return True, winner
	elif board[4] == board[5] == board[6]:
		winner = board[4]
		return True, winner
	elif board[7] == board[8] == board[9]:
		winner = board[7]
		return True, winner
	elif board[1] == board[4] == board[7]:
		winner = board[1]
		return True, winner
	elif board[2] == board[5] == board[8]:
		winner = board[2]
		return True, winner
	elif board[3] == board[6] == board[9]:
		winner = board[3]
		return True, winner
	elif board[1] == board[5] == board[9]:
		winner = board[1]
		return True, winner
	elif board[3] == board[5] == board[7]:
		winner = board[3]
		return True, winner
	elif get_valid_moves(board) == []:
		winner = None
		return True, winner
	else:
		winner = None
		return False, winner


# print_board prints the current state of the game board
def print_board(board):
	b = '''
	|%s|%s|%s|
	--+-+--
	|%s|%s|%s|
	--+-+--
	|%s|%s|%s|
	'''%(board[1], board[2], board[3], board[4], board[5], board[6], board[7], board[8], board[9])
	print(b)


# get_valid_moves returns a list of all valid moves for the current board state
def get_valid_moves(board):
	valid_moves = []
	for i in board:
		if board[i] == 'X' or board[i] == 'O':
			pass
		else:
			valid_moves.append(i)
	return valid_moves


# check_move checks whether the move made by the player is valid
def check_move(move, board):
	if move in get_valid_moves(board):
		return True
	else:
		return False


# play function is the function that actually executes the move entered by the human and the one made by the ai
def play(is_human, human, ai, board):
	if is_human == True:
		try:
			move = int(input('where do you want to play %s? '%human))
			if check_move(move, board) == True:
				board[move] = human
				print_board(board)
				print('valid moves:', get_valid_moves(board))
			else:
				print('invalid move')
				# print('invalid move: because check_move() == False')
				play(is_human, human, ai, board)
		except:
			print("invalid move")
			# print("invalid move: from play() exeption handling")
			play(is_human, human, ai, board)
	else:
	# random move:
		# board[random.choice(get_valid_moves(board))] = ai

	# best move found using minimax:
		# WARNING: i havent updated the minimax function. it WILL break things if uncommented.
		# best_score = -math.inf
		# best_move = random.choice(get_valid_moves(board))
	
		# for i in get_valid_moves(board):
		# 	# make_move(mover, i)
		# 	board[i] = mover
		# 	score = minimax(mover, board, False)
		# 	# make_move(i, i)
		# 	board[i] = i
		# 	if score > best_score:
		# 		best_score = score
		# 		best_move = i
		# board[best_move] = mover
		# print_board(board)
		# print('valid moves:', get_valid_moves(board))

	# best move found using minimax and alpha beta pruning:
		best_score = -math.inf
		best_move = random.choice(get_valid_moves(board))
	
		for i in get_valid_moves(board):
			board[i] = ai
			score = minimax_abp(ai, human, board, False, -math.inf, math.inf)
			board[i] = i
			if score > best_score:
				best_score = score
				best_move = i
		board[best_move] = ai
		print_board(board)
		print('the ai played', ai, 'at', best_move)
		print('valid moves:', get_valid_moves(board))


# game loop
def start_game(human, ai, board):
	while game_over(board)[0] == False:
		if human == 'X':
			play(True, human, ai, board)
			if game_over(board)[0] == True:
				break
			play(False, human, ai, board)
			if game_over(board)[0] == True:
				break
		else:
			play(False, human, ai, board)
			if game_over(board)[0] == True:
				break
			play(True, human, ai, board)
			if game_over(board)[0] == True:
				break
	if game_over(board)[1] == None:
		print('draw')
	else:
		if ai == game_over(board)[1]:
			print('the ai won')
		else:
			print('THIS WASNT SUPPOSED TO HAPPEN. CONTACT ME ASAP.')

# WARNING: The minimax function hasent been updated. it WILL break things if uncommented. use minimax_abp() instead.
# def minimax(mover, board, ismax):
	# if mover == 'X':
	# 	opp = 'O'
	# else:
	# 	opp = 'X'

	# if game_over(board)[0] == True:
	# 	if game_over(board)[1] == None:
	# 		return 0
	# 	elif game_over(board)[1] == mover:
	# 		return 1
	# 	else:
	# 		return -1
	# else:
	# 	if ismax == True:
	# 		best_score = -math.inf
			
	# 		for i in get_valid_moves(board):
	# 			# make_move(mover, i)
	# 			board[i] = mover
	# 			score = minimax(mover, board, False)
	# 			# make_move(i, i)
	# 			board[i] = i
	# 			if score > best_score:
	# 				best_score = score
	# 		return best_score
	# 		# best_score = max(best_score, score)
	# 		# return max(best_score, score)

	# 	else:
	# 		best_score = math.inf
			
	# 		for i in get_valid_moves(board):
	# 			# make_move(opp, i)
	# 			board[i] = opp
	# 			score = minimax(mover, board, True)
	# 			# make_move(i, i)
	# 			board[i] = i
	# 			if score < best_score:
	# 				best_score = score
	# 		return best_score
	# 		# best_score = min(best_score, score)
	# 		# return min(best_score, score)


# minimax alogrithm with alpha beta pruning
def minimax_abp(ai, human, board, is_maximizing, alpha, beta):
	# exit conditions for this recursive function
	if game_over(board)[0] == True:
		if game_over(board)[1] == None:
			return 0
		elif game_over(board)[1] == ai:
			return 1
		else:
			return -1
	# main algorithm
	else:
		if is_maximizing == True:
			best_score = -math.inf
			
			for i in get_valid_moves(board):
				board[i] = ai
				# only print the board if you want to see every move tried. might take time.
				# print_board(board)
				score = minimax_abp(ai, human, board, False, alpha, beta)
				board[i] = i
				if score > best_score:
					best_score = score
				if best_score > alpha:
					alpha = best_score
				if alpha >= beta:
					break
			return best_score
		else:
			best_score = math.inf
			
			for i in get_valid_moves(board):
				board[i] = human
				# only print the board if you want to see every move tried. might take time.
				# print_board(board)
				score = minimax_abp(ai, human, board, True, alpha, beta)
				board[i] = i
				if score < best_score:
					best_score = score
				if best_score < beta:
					beta = best_score
				if alpha >= beta:
					break
			return best_score


# main function
def main():
	# dictionary 'board' represents the main tictactoe board 
	board = {
		1: 1,
		2: 2,
		3: 3,
		4: 4,
		5: 5,
		6: 6,
		7: 7,
		8: 8,
		9: 9
	}

	# 'X' and 'O' are randomly assigned and the game will be started
	options = ('X', 'O')
	human = random.choice(options)

	if human == 'X':
		ai = 'O'
	else:
		ai = 'X'

	print('you will be playing as', human)
	print_board(board)
	print('valid moves:', get_valid_moves(board))
	start_game(human, ai, board)


#endregion


if __name__ == '__main__':
	main()

File: bot.c
#include <stdio.h>
#include <stdlib.h>
#include <limits.h>

int *GameInit(void);
void PrintBoard(int board[]);
int GetGameStatus(int board[]);
int *GetValidMoves(int board[]);
void StarGame(int board[]);
void Play(int board[], int is_ai_move);
int Minimax(int board[], int is_maximizing, int alpha, int beta);

int main(void)
{
	int *board;
	board = GameInit();
	PrintBoard(board);
	StarGame(board);
	return 0;
}

int *GameInit(void)
{
	// board[0] ==> status of game wrt x
		// 0 ==> ai playing as o
		// 1 ==> ai playing as x
	// return vals:
		// 0 ==> draw
		// 1 ==> x (ai) won
		// -1 == x (human) won
	int *board = calloc(10, sizeof(*board));
	for (int i = 1; i < 10; i++)
	{
		board[i] = i;
	}
	
	while (1)
	{
		printf("what do you wnat to play as? [x/o]\n> ");
		char choice = getchar();
		if (choice == 'x')	// ai ==> o 
		{
			board[0] = 'o' - '0';
			int _; while ((_ = getchar()) != '\n' && _ != EOF) {}	/* clear input stream */
			break;
		}
		else if (choice == 'o')	// ai ==> x
		{
			board[0] = 'x' - '0';
			int _; while ((_ = getchar()) != '\n' && _ != EOF) {}	/* clear input stream */
			break;
		}
		else
		{
			printf("\ninvalid choice! ");
			int _; while ((_ = getchar()) != '\n' && _ != EOF) {}	/* clear input stream */
		}
	}

	return(board);
}

void PrintBoard(int board[])
{
	printf("\n");
	for (int i = 1;  i < 10; i += 3)
	{
		printf("|%c|%c|%c|\n", board[i] + '0', board[i+1] + '0', board[i+2] + '0');
		if (i < 7) {printf("--+-+--\n");}
	}
	printf("\n");
}

int GetGameStatus(int board[])
{
	int *valid_moves;
	valid_moves = GetValidMoves(board);

	// board[0] ==> status of game wrt x
		// 0 ==> ai playing as o
		// 1 ==> ai playing as x
	// return vals:
		// 0 ==> draw
		// 1 ==> x (ai) won
		// -1 ==> x (human) won
		// 100 ==> game not over
	if (board[1] == board[2] && board[2] == board[3]){return (board[0] == board[1] ? 1 : -1);}
	else if (board[4] == board[5] && board[5] == board[6]){return (board[0] == board[4] ? 1 : -1);}
	else if (board[7] == board[8] && board[8] == board[9]){return (board[0] == board[7] ? 1 : -1);}
	else if (board[1] == board[4] && board[4] == board[7]){return (board[0] == board[1] ? 1 : -1);}
	else if (board[2] == board[5] && board[5] == board[8]){return (board[0] == board[2] ? 1 : -1);}
	else if (board[7] == board[6] && board[6] == board[9]){return (board[0] == board[7] ? 1 : -1);}
	else if (board[1] == board[5] && board[5] == board[9]){return (board[0] == board[1] ? 1 : -1);}
	else if (board[3] == board[5] && board[5] == board[7]){return (board[0] == board[3] ? 1 : -1);}
	else if (valid_moves[0] == 0) {return 0;}
	else {return(100);}
}

int *GetValidMoves(int board[])
{
	int *valid_moves = calloc(9, sizeof(*valid_moves));
	int valid_move_counter = 0;
	for (int i = 1; i < 10; i ++)
	{
		if (board[i] <= 9)
		{
			valid_moves[valid_move_counter] = i;
			valid_move_counter++;
		}
	}
	return (valid_moves);
}

void StarGame(int board[])
{
	int is_ai_move = (board[0] == 'x' - '0') ? 1 : 0;

	while(GetGameStatus(board) == 100)
	{
		Play(board, is_ai_move);
		is_ai_move = is_ai_move ? 0 : 1;
	}
}

void Play(int board[], int is_ai_move)
{
	if (is_ai_move)
	{
		int *valid_moves, score, best_score, best_move;
		best_score = INT_MIN;
		valid_moves = GetValidMoves(board);

		for (int i = 0; i < 9; i++)
		{
			if (valid_moves[i])
			{
				board[valid_moves[i]] = board[0];
				score = Minimax(board, 0, INT_MIN, INT_MAX);
				board[valid_moves[i]] = valid_moves[i];
				if (score > best_score)
				{
					best_score = score;
					best_move = valid_moves[i];
				}
			}
		}
		board[best_move] = board[0];
		printf("\nai played at %d\n", best_move);
		PrintBoard(board);
	}
	else
	{
		int *valid_moves;
		valid_moves = GetValidMoves(board);
		while (1)
		{
			printf("your move?\n> ");
			char choice = getchar() - '0';
			for (int i = 0; i < 9; i++)
			{
				// printf("%d, %d\n", valid_moves[i], choice);
				if (valid_moves[i] == choice)
				{
					board[valid_moves[i]] = (board[0] == 'x' - '0') ? ('o' - '0') : ('x' - '0');
					int _; while ((_ = getchar()) != '\n' && _ != EOF) {}	/* clear input stream */
					goto end;
				}
			}
			printf("\ninvalid choice! ");
			int _; while ((_ = getchar()) != '\n' && _ != EOF) {}	/* clear input stream */
		}
		end:
		PrintBoard(board);
	}
}

int Minimax(int board[], int is_maximizing, int alpha, int beta)
{
	if (GetGameStatus(board) != 100)
	{
		return GetGameStatus(board);
	}
	else
	{
		if (is_maximizing)
		{
			int *valid_moves, score, best_score, best_move;
			best_score = INT_MIN;
			valid_moves = GetValidMoves(board);
			for (int i = 0; i < 9; i++)
			{
				if (valid_moves[i])
				{
					board[valid_moves[i]] = board[0];
					score = Minimax(board, 0, INT_MIN, INT_MAX);
					board[valid_moves[i]] = valid_moves[i];
					if (score > best_score)
					{
						best_score = score;
					}
					if (best_score > alpha)
					{
						alpha = best_score;
					}
					if (alpha >= beta)
					{
						break;
					}
				}
			}
			return best_score;
		}
		else
		{
			int *valid_moves, score, best_score, best_move;
			best_score = INT_MAX;
			valid_moves = GetValidMoves(board);
			for (int i = 0; i < 9; i++)
			{
				if (valid_moves[i])
				{
					board[valid_moves[i]] = (board[0] == 'x' - '0') ? ('o' - '0') : ('x' - '0');
					score = Minimax(board, 1, INT_MIN, INT_MAX);
					board[valid_moves[i]] = valid_moves[i];
					if (score < best_score)
					{
						best_score = score;
					}
					if (best_score < beta)
					{
						beta = best_score;
					}
					if (alpha >= beta)
					{
						break;
					}
				}
			}
			return best_score;
		}
	}
}